src/celeritas/optical/detail/PreGenExecutor.hh

```diff
@@ -0,0 +1,103 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/optical/detail/PreGenExecutor.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "corecel/Macros.hh"
+#include "corecel/Types.hh"
+#include "celeritas/global/CoreTrackView.hh"
+#include "celeritas/optical/CerenkovPreGenerator.hh"
+#include "celeritas/optical/OpticalGenData.hh"
+#include "celeritas/optical/ScintillationPreGenerator.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+// LAUNCHER
+//---------------------------------------------------------------------------//
+/*!
+ * Generate optical distribution data.
+ */
+struct PreGenExecutor
+{
+    inline CELER_FUNCTION void
+    operator()(celeritas::CoreTrackView const& track);
+
+    NativeCRef<OpticalPropertyData> const properties;
+    NativeCRef<CerenkovData> const cerenkov;
+    NativeCRef<ScintillationData> const scintillation;
+    NativeRef<OpticalGenStateData> const state;
+    OpticalBufferSize size;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Generate optical distribution data.
+ */
+CELER_FUNCTION void PreGenExecutor::operator()(CoreTrackView const& track)
+{
+    CELER_EXPECT(state);
+
+    using DistId = ItemId<OpticalDistributionData>;
+
+    auto tsid = track.track_slot_id();
+    auto& cerenkov_dist = state.cerenkov[DistId(size.cerenkov + tsid.get())];
+    auto& scintillation_dist
+        = state.scintillation[DistId(size.scintillation + tsid.get())];
+
+    // Clear distribution data
+    cerenkov_dist = {};
+    scintillation_dist = {};
+
+    auto sim = track.make_sim_view();
+    bool inactive = sim.status() == TrackStatus::inactive;
+    auto optmat_id = inactive ? OpticalMaterialId{}
+                              : track.make_material_view()
+                                    .make_material_view()
+                                    .optical_material_id();
```

Comment:
@amandalund I know this has been moved around (moved to `Utils.hh` in #1217) but wanted to revisit this on the original implementation... Since the `PreGenExecutor` is a "post-post-step" action, the material here is the *post-boundary-crossing* material if a track is stopped by a boundary.  I don't think that's right... perhaps we should be collecting and storing the beginning-of-step optical material in `OpticalPreStepData`?

Reply:
Ah good catch, yeah we can store it with the rest of the pre-step data.

Reply:
@amandalund to be clear (and avoid unintentionally duplicating fixes as we've managed to do far too often 😅) would you like to address this or shall I?

Reply:
I'm happy to, if you haven't already done it 😉